ui/src/main/java/org/jtaccuino/core/ui/completion/CompletionPopup.java

```diff
@@ -26,52 +28,86 @@
 import javafx.event.EventHandler;
 import javafx.event.EventType;
 import javafx.geometry.Point2D;
-import javafx.geometry.Rectangle2D;
 import javafx.scene.Node;
 import javafx.scene.control.PopupControl;
 import javafx.scene.control.Skin;
+import javafx.stage.Window;
 
 public class CompletionPopup extends PopupControl {
 
-    private final ObservableList<String> completionSuggestions = FXCollections.observableArrayList();
+    private final ObservableList<CompletionItem> completionSuggestions = FXCollections.observableArrayList();
     private IntegerProperty visibleCompletions = new SimpleIntegerProperty(this, "visibleCompletions", 10);
 
-    @SuppressWarnings("serial")
     public static class CompletionEvent extends Event {
 
+        @java.io.Serial
+        static final long serialVersionUID = -1L;
+
         public static final EventType<CompletionEvent> COMPLETION
                 = new EventType<>("Completion" + UUID.randomUUID().toString()); //$NON-NLS-1$
 
         private final String suggestion;
+        private final int anchor;
 
-        public CompletionEvent(String suggestion) {
+        public CompletionEvent(String suggestion, int anchor) {
             super(COMPLETION);
             this.suggestion = suggestion;
+            this.anchor = anchor;
         }
 
         public String getSuggestion() {
             return suggestion;
         }
+
+        public int getAnchor() {
+            return anchor;
+        }
     }
 
     @SuppressWarnings("this-escape")
     public CompletionPopup() {
         setAutoFix(true);
-        setAutoHide(true);
+        setAutoHide(false);
         setHideOnEscape(true);
     }
 
-    public ObservableList<String> getSuggestions() {
+    public ObservableList<CompletionItem> getSuggestions() {
         return completionSuggestions;
     }
 
+    @SuppressWarnings("CatchAndPrintStackTrace")
+    public void setSuggestions(List<CompletionItem> items) {
+        var matchList = items.stream().filter(CompletionItem::matchesType).toList();
+        var noMatchlist = items.stream().filter(CompletionItem::notMatchesType).toList();
+        boolean requiresSep = !matchList.isEmpty() && !noMatchlist.isEmpty();
+        @SuppressWarnings("unchecked")
+        var newList = FXCollections.concat(
+                FXCollections.observableArrayList(matchList),
+                requiresSep ? FXCollections.observableArrayList(CompletionItem.NIL) : FXCollections.emptyObservableList(),
+                FXCollections.observableArrayList(noMatchlist));
```

Comment:
I'm guessing the warning comes from the emptyObservableList call. Giving that call the actual type should fix it

```suggestion
        var newList = FXCollections.concat(
                FXCollections.observableArrayList(matchList),
                requiresSep ? FXCollections.observableArrayList(CompletionItem.NIL) : FXCollections.<CompletionItem>emptyObservableList(),
                FXCollections.observableArrayList(noMatchlist));
```